src/sc/sidechain.cpp

```diff
@@ -757,61 +778,139 @@ void CSidechain::UpdateScFees(const CScCertificateView& certView)
+    // v1 sidechains
+    if (!isNonCeasing()) {
+        scFees.push_back(Sidechain::ScFeeData(ftScFee, mbtrScFee));
 
-    // check if we are past the buffer size
-    int delta = scFees.size() - maxSizeOfScFeesContainers;
-    if (delta > 0)
-    {
-        // remove from the front as many elements are needed to be within the circular buffer size
-        // --
-        // usually this is just one element, but in regtest a node can set the max size via a startup option
-        // therefore such size might be lesser than scFee size after a node restart
-        while (delta > 0)
+        int delta = scFees.size() - maxSizeOfScFeesContainers;
+        if (delta > 0)
         {
-            if (scFees.empty())
-                break;
+            // remove from the front as many elements are needed to be within the circular buffer size
+            // --
+            // usually this is just one element, but in regtest a node can set the max size via a startup option
+            // therefore such size might be lesser than scFee size after a node restart
+            while (delta > 0)
+            {
+                if (scFees.empty())
+                    break;
+
+                const auto& entry = scFees.front();
+                LogPrint("sc", "%s():%d - popping f=%d, m=%d from list, as scFees maxsize has been reached\n",
+                    __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
+                scFees.pop_front();
+                delta--;
+            }
+        }
+    }
+    // v2 sidechains
+    else {
+        auto scFeeIt = std::find_if(scFees_v2.begin(), scFees_v2.end(), 
+            [&blockHeight](Sidechain::ScFeeData_v2 scFeeElem){return scFeeElem.submissionHeight == blockHeight;});
 
-            const auto& entry = scFees.front();
-            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list\n",
-                __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
-            scFees.pop_front();
-            delta--;
+        // We have not found a scFeeData for the current height, so we add a new one and perform all the checks
+        // on the container size / element age
+        if (scFeeIt == scFees_v2.end()) {
+            scFees_v2.push_back(Sidechain::ScFeeData_v2(ftScFee, mbtrScFee, blockHeight));
+
+            // as for v1 sidechains
+            int delta = scFees_v2.size() - maxSizeOfScFeesContainers;
+            if (delta > 0)
+            {
+                while (delta > 0)
+                {
+                    if (scFees_v2.empty())
+                        break;
+
+                    const auto& entry = scFees_v2.front();
+                    LogPrint("sc", "%s():%d - popping f=%d, m=%d, h=%d from list, as scFees maxsize has been reached\n",
+                        __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee, entry.submissionHeight);
+                    scFees_v2.pop_front();
+                    delta--;
+                }
+            }
```

Comment:
The `if (delta > 0)` is not needed here imho. Having no `else` branch the `while (delta > 0)` is enough

Reply:
fixed in f0bab99cd7559a9bab4c8dc65103df6eea2b4b76

Reply:
Yet a small note on this
In both loops I see 

```
        int delta = scFees.size() - maxSizeOfScFeesContainers;
        while (delta > 0)
        {
            if (scFees.empty())
                break;

            const auto& entry = scFees.front();
            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list, as scFees maxsize has been reached\n",
                __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
            scFees.pop_front();
            --delta;
        }
```

As a result the inner check `if (scFees.empty())` can really never happen : we enter the while loop only if vector's size is >  maxSizeOfScFeesContainers and by consequence we never empty the vector.
As a suggestion I'd rewrite as : 

```
        const size_t max_size { static_cast<size_t>(maxSizeOfScFeesContainers) };
        while (scFees.size() > max_size)
        {
            const auto& entry = scFees.front();
            LogPrint("sc", "%s():%d - popping f=%d, m=%d from list, as scFees maxsize has been reached\n",
                __func__, __LINE__, entry.forwardTxScFee, entry.mbtrTxScFee);
            scFees.pop_front();
        }

```

Which is more readable and clearly expresses the intention without delta and substractions